modulemd/modulemd-packager-v3.c

```diff
@@ -983,6 +983,198 @@ modulemd_packager_v3_to_stream_v2_ext (ModulemdPackagerV3 *self,
+      return NULL;
+    }
+
+  /* create a ModuleIndex to contain the results */
+  index = modulemd_module_index_new ();
+
+  /* create a StreamV2 object for each BuildConfig object */
+  for (guint i = 0; i < g_strv_length (contexts); i++)
+    {
+      bc = modulemd_packager_v3_get_build_config (self, contexts[i]);
+
+      v2_stream = modulemd_module_stream_v2_new (
+        modulemd_packager_v3_get_module_name (self),
+        modulemd_packager_v3_get_stream_name (self));
+
+      modulemd_module_stream_set_context (MODULEMD_MODULE_STREAM (v2_stream),
+                                          contexts[i]);
+
+      /* set attributes that are the same for all streams */
+
+      modulemd_module_stream_v2_set_summary (
+        v2_stream, modulemd_packager_v3_get_summary (self));
+
+      modulemd_module_stream_v2_set_description (
+        v2_stream, modulemd_packager_v3_get_description (self));
+
+      /* Packager v3 "license" is optional. Stream v2 "license" is required
+       * Fill in the default Packager v3 license if none has been specified.
+       */
+      if (g_hash_table_size (self->module_licenses) == 0)
+        {
+          modulemd_module_stream_v2_add_module_license (
+            v2_stream, MMD_PACKAGER_DEFAULT_MODULE_LICENSE);
+        }
+      else
+        {
+          MODULEMD_REPLACE_SET (v2_stream->module_licenses,
+                                self->module_licenses);
+        }
+
+      modulemd_module_stream_v2_set_xmd (v2_stream,
+                                         modulemd_packager_v3_get_xmd (self));
+
+      modulemd_module_stream_v2_set_community (
+        v2_stream, modulemd_packager_v3_get_community (self));
+
+      modulemd_module_stream_v2_set_documentation (
+        v2_stream, modulemd_packager_v3_get_documentation (self));
+
+      modulemd_module_stream_v2_set_tracker (
+        v2_stream, modulemd_packager_v3_get_tracker (self));
+
+      g_hash_table_iter_init (&iter, self->profiles);
+      while (g_hash_table_iter_next (&iter, NULL, &value))
+        {
+          profile = modulemd_profile_copy (MODULEMD_PROFILE (value));
+          modulemd_profile_unset_default (profile);
+          modulemd_module_stream_v2_add_profile (v2_stream, profile);
+          g_clear_object (&profile);
+        }
+
+      modulemd_module_stream_v2_replace_rpm_api (v2_stream, self->rpm_api);
+
+      modulemd_module_stream_v2_replace_rpm_filters (v2_stream,
+                                                     self->rpm_filters);
+
+      COPY_HASHTABLE_BY_VALUE_ADDER (v2_stream,
+                                     self,
+                                     rpm_components,
+                                     modulemd_module_stream_v2_add_component);
+
+      COPY_HASHTABLE_BY_VALUE_ADDER (v2_stream,
+                                     self,
+                                     module_components,
+                                     modulemd_module_stream_v2_add_component);
```

Comment:
This chunk of code is probably going to need to be reused for the `ModuleStream.read_*()` codepath, so it is probably best to extract it into its own function. Probably not a terrible idea to reuse it for `ModuleStreamV2.copy()` as well. Then we only have to worry about a single place to update in the future.

Reply:
Redundancy eliminated for both chunks. But I didn't attempt to re-use it for `ModuleStreamV2.copy()` due to the incompatible object types.